2 Niagara Fundamentals Exercise - Multiplexer/multiplexer-rt/src/com/tridiumuniversity/multiplexer/BMultiplexer.java

```diff
@@ -0,0 +1,284 @@
+        SlotCursor<Property> propertyCursor = getProperties(); // create the cursor
+
+        while (propertyCursor.next()) // iterate through the properties
+        {
+            Property property = propertyCursor.property();
+            try
+            {
+                BBoolean value = (BBoolean) propertyCursor.get(); // get property value. If the property isn't a BBoolean, skip
+
+                if (property.getName().startsWith("in")) {
+                    inInputCount += 1; // add one to the count
+
+                    inInputs.add(property);
+                }
+            }
+            catch (Exception e)
+            {
+            }
+        }
+        log.fine("Counted INs: " + inInputCount); // FOR SOME REASON THIS LOGS IN THE WB CONSOLE WHEN SWITCHING VIEWS
+    }
+
+
+    /* Calculates how many "s" properties exist in the multiplexer */
+    private void calcSInputCount()
+    {
+        sInputCount = 0; // reset the old count
+        sInputs.clear(); // clear the old "s" input list
+        SlotCursor<Property> propertyCursor = getProperties(); // create the cursor
+
+        while (propertyCursor.next()) // iterate through the properties
+        {
+            Property property = propertyCursor.property(); // get property object
+            try
+            {
+                BBoolean value = (BBoolean) propertyCursor.get(); // get property value. If the property isn't a BBoolean, skip
+
+                if (property.getName().startsWith("s")) {
+                    sInputCount += 1; // add one to the count
+
+                    sInputs.add(property);
+                }
+            }
+            catch (Exception e)
+            {
+            }
+        }
+        log.fine("Counted Ss: " + sInputCount); // FOR SOME REASON THIS LOGS IN THE WB CONSOLE WHEN SWITCHING VIEWS
+    }
+
+
+    /* Calculates the effective input step. The input step decides how many sInputs are needed/used */
+    private void calcInputStep()
+    {
+        /* -1 in the loop so the iteration number matches with the index number of the array.
+        I don't know if this is how it's normally done. It's the first time I've ever wanted to count backwards in a loop */
+        for (int i = inputSteps.length - 1; i>-1; i--)
+        //
+        {
+            if (inInputCount >= inputSteps[i])
+            {
+                effInputStep = i;
+                log.fine("Effective input step: " + effInputStep);
+                break;
+            }
+        }
+    }
+
+
+    /* Calculates the switchSum based on the combination of switches currently set to "true" */
+    private void calcSwitchSum()
+    {
+        switchSum = 0; // reset the switchSum
+
+        /* process all the "in" inputs */
+        for (int i = 0; i<sInputs.size(); i++)
+        {
+            if (((BBoolean) sInputs.get(i).getDefaultValue()).getBoolean()) // get input as boolean. If true, add a numeric value to the switchSum
```

Comment:
We have several getter methods defined on `BComplex` that could make this a bit simpler. I might write this line as:

```java
if(getBoolean(sInputs.get(i)))
```